packages/wrangler/src/cloudchamber/apply.ts

```diff
@@ -0,0 +1,566 @@
+		return object.map((obj) => sortObjectRecursive(obj)) as T;
+	}
+
+	const objectCopy: Record<string | number, unknown> = { ...object };
+	for (const [key, value] of Object.entries(object)) {
+		if (typeof value === "object") {
+			if (value === null) {
+				continue;
+			}
+			objectCopy[key] = sortObjectRecursive(
+				value as Record<string, unknown>
+			) as unknown;
+		}
+	}
+
+	return sortObjectKeys(objectCopy) as T;
+}
+
+/**
+ * applyCommand is able to take the wrangler.toml file and render the changes that it
+ * detects.
+ */
+export async function applyCommand(
+	args: StrictYargsOptionsToInterfaceJSON<typeof applyCommandOptionalYargs>,
+	config: Config
+) {
+	startSection(
+		"Deploy a container application",
+		"deploy changes to your application"
+	);
+
+	if (config.container_app.length === 0) {
+		endSection(
+			"You don't have any container applications defined in your wrangler.toml",
+			"You can set the following configuration in your wrangler.toml"
+		);
+		const configuration = {
+			configuration: {
+				image: "docker.io/cloudflare/hello-world:1.0",
+			},
+			instances: 2,
+			name: config.name ?? "my-containers-application",
+		};
+		const endConfig: JsonMap =
+			args.env !== undefined
+				? {
+						env: { [args.env]: { container_app: [configuration] } },
+					}
+				: { container_app: [configuration] };
+		TOML.stringify(endConfig)
+			.split("\n")
+			.map((el) => el.trim())
+			.forEach((el) => {
+				printLine(el, "  ", logRaw);
+			});
+		return;
+	}
+
+	const applications = await promiseSpinner(
+		ApplicationsService.listApplications(),
+		{ json: args.json, message: "Loading applications" }
+	);
+	const applicationByNames: Record<ApplicationName, Application> = {};
+	// TODO: this is not correct right now as there can be multiple applications
+	// with the same name.
+	for (const application of applications) {
+		applicationByNames[application.name] = application;
+	}
+
+	const actions: (
+		| { action: "create"; application: CreateApplicationRequest }
+		| {
+				action: "modify";
+				application: ModifyApplicationRequestBody;
+				id: ApplicationID;
+				name: ApplicationName;
+		  }
+	)[] = [];
+
+	log(dim("Container application changes\n"));
+
+	for (const appConfigNoDefaults of config.container_app) {
+		const appConfig = containerAppToCreateApplication(
+			appConfigNoDefaults,
+			args.skipDefaults
+		);
+		const application = applicationByNames[appConfig.name];
+		if (application !== undefined && application !== null) {
+			// we need to sort the objects (by key) because the diff algorithm works with
+			// lines
+			const prevApp = sortObjectRecursive<CreateApplicationRequest>(
+				stripUndefined(applicationToCreateApplication(application))
+			);
+
+			const prev = TOML.stringify({ container_app: [prevApp] });
+			const now = TOML.stringify({
+				container_app: [
+					sortObjectRecursive<CreateApplicationRequest>(appConfig),
+				],
+			});
+			const results = Diff.diffLines(prev, now);
```

Comment:
This seems slightly overkill, and I'm a bit concerned about adding a ~600kb dependency just for this. What type of data is this likely to be? If it's an objet, have you considered just diffing by key/value? That feels like it'd be sufficient to me, and wouldn't need a text diffing algorithm

Reply:
I can take a look at the dependency we are using and see if we can simplify, as 600kb for line diffing is too much. However, I'd argue that line diffing is the right approach here because we have complex nested objects inside arrays etc. that will make obj diffing approach a bit of a mess to test and render, and I've observed that line diffing to be easier to reason about once you just order the object properties.